llvm/tools/sycl-post-link/SpecConstants.h

```diff
@@ -34,9 +43,13 @@ class SpecConstantsPass : public PassInfoMixin<SpecConstantsPass> {
 
   // Searches given module for occurences of specialization constant-specific
   // metadata at call instructions and builds a
-  // "spec constant name" -> "spec constant int ID" map from this information.
+  // "spec constant name" -> "spec constant int ID" map for scalar spec
+  // constants and
+  // "spec constant name" -> vector<"spec constant int ID"> map for composite
+  // spec constants
```

Comment:
What is the reason for two different maps for scalars and composites?

Reply:
Because structures representing scalar and composite spec constants are different, i.e. for scalar spec constant we only need to have a mapping from string to int, while for composite spec constants we need to have a mapping from string to vector.

It is easier to maintain two maps because those maps are stored into separate properties of device image

Reply:
Can we add mapping for scalar spec constants from string to vector with one element?

> It is easier to maintain two maps because those maps are stored into separate properties of device image

Why it is required to store separate properties for scalar and composite spec constants?

Reply:
> Can we add mapping for scalar spec constants from string to vector with one element?

Technically this is double, but there is a significant downside which I will describe below:

> Why it is required to store separate properties for scalar and composite spec constants?

It is easier and probably more clean and reliable way of differentiating them in runtime: instead of checking for the size of vector I just scan two different properties in order to handle scalar and composite spec constants and their handling is different there

Reply:
> It is easier and probably more clean and reliable way of differentiating them in runtime: instead of checking for the size of vector I just scan two different properties in order to handle scalar and composite spec constants and their handling is different there

I guess their handling only is different because you need to set several values of spec constants (I mean do several calls of smth like clSetSpecConst or how it is called, I don't know exactly) in case of composite spec constant and only one in case of scalar spec constant. I believe if we save scalar constants as vector of one element and handle it in the same way as it was scalar spec constant, you won't need to differentiate them in RT and sycl-post-link. Although this is probably a big change across two components and can be done in the future as some kind of refactoring.

Reply:
>  I believe if we save scalar constants as vector of one element and handle it in the same way as it was scalar spec constant, you won't need to differentiate them in RT and sycl-post-link.

This is a good point, actually, thanks!

>  Although this is probably a big change across two components and can be done in the future as some kind of refactoring.

Yes, we would also need to update our design document to unify representation/handling of scalar and composite spec constants. I would like to hear feedback from @kbobrovs about this before proceeding with this refactoring

Reply:
This makes sense to me. But I'm assuming SPIRV and __sycl intrinsics will remain separate.